### PR TITLE
record/class reading and JSON improvements

### PR DESCRIPTION
**Pull request by mppf**
This PR includes:

* style element to skip unknown fields
* writef/readf format string: ~ means skip unknown fields
* new channel.skipField including C implementation for JSON
* improvements to the default-generated record/class reader
  * reads fields in any order
  * fixes bugs when reading classes with inheritance
  * supports QIO_STYLE_ELEMENT_SKIP_UNKNOWN_FIELDS
* Several improvements to List.chpl, including support for JSON format
* Tests demonstrating the new features

Passed full local testing.
Passed GASNet+numa testing on test/io/ferguson.
Reviewed by @lydia-duncan - thanks!

### CODE REVIEW

File: modules/internal/ChapelIO.chpl [hunk 1]
```diff
@@ -345,50 +349,225 @@ module ChapelIO {
...
+            // the field name
+            var fname:ioLiteral;
+
+            var read_field_name = false;
+
+            //writeln("BEFORE PARAM FOR type ", t:string, " num_fields=", num_fields, " read_field_name=", read_field_name);
+
+            for param i in 1..num_fields {
+              if isType(__primitive("field value by num", x, i)) ||
+                 isParam(__primitive("field value by num", x, i)) {
```
Comment (lydia-duncan) [inline modules/internal/ChapelIO.chpl]: It might be worthwhile to make this pair of checks into a single call so you can reuse it.

File: modules/internal/ChapelIO.chpl [hunk 2]
```diff
@@ -345,50 +349,225 @@ module ChapelIO {
...
+                // break out of the loop if we didn't read a comma
+                // and we're expecting to read one.
+                break;
+              }
+            }
+
+            // find a field name that matches.
+            // TODO: this is not particularly efficient. If we
+            // have a lot of fields, this is O(n**2), and there
+            // are other potential problems with string reallocation.
```
Comment (lydia-duncan) [inline modules/internal/ChapelIO.chpl]: Do you have ideas on how to make it more efficient, or is this just a reminder to think of them?

File: modules/internal/ChapelIO.chpl [hunk 3]
```diff
@@ -345,50 +349,225 @@ module ChapelIO {
...
+                reader.skipField(error=e);
+                if !e {
+                  needsComma = true;
+                }
+                reader.setError(e);
+              } else {
+                //writeln("ERROR - MISSING FIELD NAME");
+                reader.setError(EFORMAT:syserr);
+                break;
+              }
```
Comment (lydia-duncan) [inline modules/internal/ChapelIO.chpl]: This logic looks very similar to that function you had earlier.  Potential for reuse?

File: modules/internal/ChapelIO.chpl [hunk 4]
```diff
@@ -345,50 +349,225 @@ module ChapelIO {
...
+            }
+          }
+
+          //writeln("POST WHILE ERROR IS ", reader.error():int);
+
+          // check that we've read all fields, return error if not.
+          {
+            var ok = num_read == num_to_read;
+            /*for param i in 1..num_fields {
```
Comment (lydia-duncan) [inline modules/internal/ChapelIO.chpl]: Leftover code to remove?

File: modules/internal/ChapelIO.chpl [hunk 5]
```diff
@@ -403,23 +582,48 @@ module ChapelIO {
           }
         } else {
           // Read the field name = part until we get one that worked.
+          var found_field = false;
           for param i in 1..num_fields {
             var st = reader.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
             var eq:ioLiteral;
```
Comment (lydia-duncan) [inline modules/internal/ChapelIO.chpl]: It looks like you've left "eq" in this scope here but you redeclare it in an inner scope.  I'm guessing that wasn't intentional?

File: modules/internal/ChapelIO.chpl (hunk 2 above)
Comment (mppf) [inline modules/internal/ChapelIO.chpl]: We could put all of the fields into a regular expression, separated by |. That would match any field. If we put each into a capture group, we could even know which field we matched.

File: modules/internal/ChapelIO.chpl (hunk 3 above)
Comment (mppf) [inline modules/internal/ChapelIO.chpl]: I actually used to have these two cases folded together, but I became much less confused about what the outer loop does when I took them apart. They have slightly different error handling. This chunk of code is short enough that it doesn't seem worth it to me to make a function that behaves differently based on an argument.

File: modules/internal/ChapelIO.chpl (hunk 2 above)
Comment (lydia-duncan) [inline modules/internal/ChapelIO.chpl]: Sorry, should have been more clear.  I'm mostly wondering on whether this should wait for the more efficient implementation to go in so that we're certain it does, or whether it is bouncing around your head enough that you wanted the initial implementation to go in and would get back to doing the improved version in short order.

File: modules/internal/ChapelIO.chpl (hunk 3 above)
Comment (lydia-duncan) [inline modules/internal/ChapelIO.chpl]: Seems fair to me!

File: modules/internal/ChapelIO.chpl (hunk 2 above)
Comment (mppf) [inline modules/internal/ChapelIO.chpl]: The use cases I've been working on have a small number of fields, so this is unlikely to be a performance problem there. I could add the regexp version but it would add complexity and make the feature dependent on having Chapel built with regexp support. I plan to leave it as-is for those reasons.

File: modules/standard/IO.chpl [hunk 6]
```diff
@@ -6348,6 +6349,35 @@ proc readf(fmt:string):bool {
...
+/*
+   Skip a field in the current aggregate format. This method is currently only
+   supported for JSON format and returns ENOTSUP for other formats. In other
+   formats, it may not be possible in general to know when a field ends.
+
+   :arg error: optional argument to capture an error code. If this argument
+               is not provided and an error is encountered, this function
+               will halt with an error message.
+ */
+proc channel.skipField(out error:syserr) {
```
Comment (lydia-duncan) [inline modules/standard/IO.chpl]: Does this skip the next field to be read?  Would it be worthwhile to provide a version where you name the field you want skipped?

File: modules/standard/List.chpl [hunk 7]
```diff
@@ -57,6 +57,13 @@ record list {
   var length: int;
 
   /*
+     Synonym for length.
+   */
+  proc size {
+    return length;
+  }
```
Comment (lydia-duncan) [inline modules/standard/List.chpl]: I'm not sure I see the value of adding this, push_back and push_front.

File: modules/standard/IO.chpl (hunk 6 above)
Comment (mppf) [inline modules/standard/IO.chpl]: This function supports the read-and-skip-unknown-fields. Suppose we had a record with a, b.
This function will skip c in this:
{"a":1, "b":2, "c":3}

But c isn't a field in our record - we don't know the name of the field we are skipping.
It skips the next field to be read. It does not include the field name.

File: modules/standard/List.chpl (hunk 7 above)
Comment (mppf) [inline modules/standard/List.chpl]: I wanted code to be able to use an array or a list without any changes, other than the data type it is working with.

File: modules/standard/IO.chpl (hunk 6 above)
Comment (mppf) [inline modules/standard/IO.chpl]: Sorry, I misspoke. It does skip the field name part also. But not e.g. a comma afterwards.

Comment (mppf) [inline modules/standard/IO.chpl]: Anyway, to answer your question, I don't have a use in mind for a version that took in the field name to skip.

File: modules/standard/List.chpl [hunk 8]
```diff
@@ -144,17 +180,121 @@ record list {
...
     for e in this {
-      if !first then
-        f.write(" ");
-      else
-        first = false;
-      f.write(e);
+      if first then first = false;
+      else {
+        if isspace then f <~> new ioLiteral(" ");
+        else if isjson || ischpl then f <~> new ioLiteral(",");
```
Comment (lydia-duncan) [inline modules/standard/List.chpl]: Should there be a space after the comma?  I think this would create [a,b,c] when we might prefer [a, b, c]

Comment (lydia-duncan) [inline modules/standard/List.chpl]: But I might be confused on the use of this section

Comment (mppf) [inline modules/standard/List.chpl]: I think you understand correctly. Do you think we should print that space by default? It is not strictly necessary.

Comment (lydia-duncan) [inline modules/standard/List.chpl]: I like the spaced out version better, but that's a personal preference.

File: modules/standard/List.chpl [hunk 9]
```diff
@@ -144,17 +180,121 @@ record list {
...
+          if f.error() == EFORMAT {
+            f.clearError();
+          } else {
+            read_end = true;
+            break;
+          }
+        } else {
+          // read a comma or a space.
+          if isspace then f <~> new ioLiteral(" ");
+          else if isjson || ischpl then f <~> new ioLiteral(",");
```
Comment (lydia-duncan) [inline modules/standard/List.chpl]: Space here, too (if you go with that, up to you)

File: runtime/src/qio/qio_formatted.c [hunk 10]
```diff
@@ -1514,6 +1514,362 @@ qioerr qio_quote_string(uint8_t string_start, uint8_t string_end, uint8_t string
 #undef WRITEC
 }
 
+static inline bool is_json_whitespace(int32_t c)
+{
+  return ( c == ' ' || c == '\b' ||
+           c == '\f' || c == '\n' || c == '\r' || c == '\t' );
+}
+
+// Read and skip an abritrary JSON object, assuming the leading '{'
```
Comment (lydia-duncan) [inline runtime/src/qio/qio_formatted.c]: "abritrary" -> "arbitrary"

File: runtime/src/qio/qio_formatted.c [hunk 11]
```diff
@@ -1514,6 +1514,362 @@ qioerr qio_quote_string(uint8_t string_start, uint8_t string_end, uint8_t string
...
+    } else if( c == '}' ) {
+      // we've reached the end.
+      return 0;
+    } else {
+      return -EFORMAT;
+    }
+  }
+}
+
+// Read and skip an abritrary JSON array, assuming the leading '['
```
Comment (lydia-duncan) [inline runtime/src/qio/qio_formatted.c]: Same spell check here

File: modules/standard/List.chpl (hunk 9 above)
Comment (mppf) [inline modules/standard/List.chpl]: In the reader, we don't want to require the space. ioLiteral defaults to consuming extra space.

Comment (lydia-duncan) [inline modules/standard/List.chpl]: Ah, okay

File: runtime/src/qio/qio_formatted.c [hunk 12]
```diff
@@ -1514,6 +1514,362 @@ qioerr qio_quote_string(uint8_t string_start, uint8_t string_end, uint8_t string
...
+    if( c == '\"' ) return 0;
+
+    // backslash: handle quoting.
+    if( c == '\\' ) {
+      c = qio_channel_read_byte(false, ch);
+      if( c < 0 ) return c;
+
+      // As long as we handle \" correctly, there is
+      // no need to validate the JSON in the string.
+      /*
```
Comment (lydia-duncan) [inline runtime/src/qio/qio_formatted.c]: Some more leftover code to clean up?

File: runtime/src/qio/qio_formatted.c [hunk 13]
```diff
@@ -3633,6 +3989,7 @@ qioerr qio_conv_parse(c_string fmt,
   int minus_flag = 0;
   int space_flag = 0;
   int plus_flag = 0;
+  int sloppy_flag = 0;
```
Comment (lydia-duncan) [inline runtime/src/qio/qio_formatted.c]: Heh.  I might call this "casual" rather than "sloppy", but either is fine.

File: test/io/ferguson/json-list2.chpl [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+var f = opentmp();
+{
+  var writer = f.writer();
+  var str = '{"numbers": []}';
+  writeln("Writing JSON: ", str);
+  writer.write(str);
+  writer.close();
+}
+
```
Comment (lydia-duncan) [inline test/io/ferguson/json-list2.chpl]: These are identical checks to the previous file.  Would it be reasonable to make a module with all of them as functions in it, which both tests could use and call?

Comment (lydia-duncan) [inline test/io/ferguson/json-list2.chpl]: At least for the shared ones

File: test/io/ferguson/read-class-reorder.chpl [hunk 15]
```diff
@@ -0,0 +1,29 @@
+/*class Base {
+}
+
+class Parent : Base {
+  var x: int;
+  var y: int;
+}
+*/
+
+class Child /*: Parent*/ {
```
Comment (lydia-duncan) [inline test/io/ferguson/read-class-reorder.chpl]: There's some leftover code here, do you want to copy it to a second test?  It seems like that would make a good edge case to me

**Comment by lydia-duncan** [comment]
It looks like there's some conflicts to resolve, and I'd recommend another sanity full local test run, but I think this is good to go otherwise.

File: test/io/ferguson/json-list2.chpl (hunk 14 above)
Comment (mppf) [inline test/io/ferguson/json-list2.chpl]: I tend to prefer simplicity in the tests over having less code. I don't think that these tests will be a maintenance burden.

File: test/io/ferguson/read-class-reorder.chpl (hunk 15 above)
Comment (mppf) [inline test/io/ferguson/read-class-reorder.chpl]: This leftover code is in read-class-inherited2 already.

Comment (lydia-duncan) [inline test/io/ferguson/read-class-reorder.chpl]: Yeah, but that one only covered inherited fields.  I was thinking of an overlap test - inherited fields plus assignment in different order.

Comment (mppf) [inline test/io/ferguson/read-class-reorder.chpl]: If I understand what you're describing - it doesn't actually work. Although we can now handle fields being read in a different order than declaration order, if there is inheritance, the parent class fields must be read before the child class fields.

Comment (lydia-duncan) [inline test/io/ferguson/read-class-reorder.chpl]: Interesting.  Is this something we'd expect/like to work down the road?  Might be worth adding a feature request/bug future.  Also, might be worthwhile to check the reordering within either parent or child fields, as opposed to across the boundaries.

Comment (mppf) [inline test/io/ferguson/read-class-reorder.chpl]: It's not so clear to me how valuable it is as a feature.... I think I'd prefer to wait for someone to say that they care about it. I've added a test for reordering within parent/child.

Comment (lydia-duncan) [inline test/io/ferguson/read-class-reorder.chpl]: I view it as a way of documenting this tension.  If we get far enough down the road that neither you nor I remember this issue when someone encounters it, it would be nice to be able to check the STATUS file and see that the behavior was at least expected, as opposed to something we hadn't even thought about.

Comment (mppf) [inline test/io/ferguson/read-class-reorder.chpl]: Fair enough. I've added such a test.
